add TPC cluster reject options

KlewinS opened the pull request:
- added the option to reject clusters:
  - single pad clusters
  - single time clusters
  - peak in the same pad but later timebin

- added first "charge sharing" option, clusters are split at the minimum, the minimum itself contributes half to all clusters.

- added option to **NOT** clear output containers by calling process/finishProcess functions (@matthiasrichter maybe this helps with the reco-workflow, could you please test it)

KlewinS commented:
@shahor02 I have split the commits and changed the `ClustererSpec.cxx` accordingly. Please have a look. In addition, I also fixed the the test, thanks for spotting it.

shahor02 commented:
@KlewinS : thanks, I've checked the cont.mode reconstruction with this PR (needed to rebase to dev), everything looks ok. Do you have other reasons to keep it is as WIP?

KlewinS commented:
no, if you checked this, I can remove the WIP.

sawenzel commented:
@wiechula : Can you approve this PR, please?

KlewinS commented:
@sawenzel  do you know by chance why the code-check fails? there are errors at the very end like

`493 warnings and 20 errors generated.
Error while processing /mnt/mesos/sandbox/sandbox/sw/SOURCES/O2/build_o2checkcode_o2/0/Detectors/TPC/reconstruction/src/HwClusterer.cxx.
error: too many errors emitted, stopping now [clang-diagnostic-error]
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/GCC-Toolchain/v7.3.0-alice1-2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:113:15: error: unknown type name 'basic_string_view' [clang-diagnostic-error]
      typedef basic_string_view<_CharT, _Traits> __sv_type;
              ^
/mnt/mesos/sandbox/sandbox/sw/slc7_x86-64/GCC-Toolchain/v7.3.0-alice1-2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/basic_string.h:113:32: error: expected member name or ';' after declaration specifiers [clang-diagnostic-error]
      typedef basic_string_view<_CharT, _Traits> __sv_type;
                               ^`

and so on. I can't see how this is linked to the PR.

KlewinS commented on Detectors/TPC/reconstruction/include/TPCReconstruction/HwClusterer.h:
```diff
@@ -163,29 +185,25 @@ class HwClusterer : public Clusterer
    * class members
    */
-  static const int mTimebinsInBuffer = 5;
+  static const int mTimebinsInBuffer = 6;
```
This change went along with the change in line 213 (and 815) `computeClusterForTime(i - 2);` -> `computeClusterForTime(i` - 3);`

I've seen, that to calculate all the relationships of all cluster pads in advance, I need the 6th timebins and have look for clusters in 1 timebin later than before (-3 instead of -2). This is only now important because possible minima of the newest (last) timebin are taken into account for the splitting. Before, this information was simply not used and therefore it did not matter that it was not yet calculated (and therefore I did not see it before).

<img width="330" alt="bildschirmfoto 2018-10-19 um 20 02 21" src="https://user-images.githubusercontent.com/15690984/47235630-ea799280-d3d9-11e8-8295-ca3da873dc28.png">

KlewinS commented:
@wiechula concerning the speed improvement of Vc: being honest, I did not do a direct comparison between a version without and a version with. There were many changes in between. I implemented that all relations between all pads are calculated and found that the speed was noticeable slower than before. Therefore I implemented the Vc and had the impression (the implementation took also some weeks) that it was maybe a factor 2 faster afterwards compared to the changed version before (or maybe not, the implementation took some weeks...).
